wgengine/magicsock/pathfinder.go

```diff
@@ -4,10 +4,101 @@
 
 package magicsock
 
-// startPathFinder initializes the atomicSendFunc, and
+import (
+	"fmt"
+	"net/netip"
+	"time"
+
+	"tailscale.com/tstime/mono"
+	"tailscale.com/types/key"
+)
+
+// startPathFinder initializes the sendFunc, and
 // will eventually kick off a goroutine that monitors whether
 // that sendFunc is still the best option for the endpoint
 // to use and adjusts accordingly.
 func (de *endpoint) startPathFinder() {
-	de.pathFinderRunning = true
+	go func() {
+		for mono.Since(de.lastSendAtomic.Load()) < sessionActiveTimeout {
+			de.pathFinderRunning = true
+
+			// while the session has not timed out yet,
+			// check whether path needs to be upgraded on an interval
+			de.updateSendPathIfNecessary(mono.Now())
+
+			// TODO(2022-10-20): should not be using heartbeat at all, currently just
+			// trying to replicate existing behaviour
+			time.Sleep(heartbeatInterval)
+		}
+	}()
+}
+
+// updateSendPathIfNecessary optionally upates sendFunc
+// based on analysis of current conditions
+func (de *endpoint) updateSendPathIfNecessary(now mono.Time) {
+	de.mu.Lock()
+	defer de.mu.Unlock()
+
+	// default happy state is: use UDP, don't use Derp
+	useUDP := true
+	useDerp := false
```

Comment:
This feels not-idiomatic, but it also seemed like the cleanest way to avoid a bunch of nested `.IsValid()` checks. Open to suggestions!